subsys/usb/class/netusb/function_rndis.c

```diff
@@ -0,0 +1,1054 @@
+		break;
+	case RNDIS_OBJECT_ID_GEN_RECEIVE_NO_BUF:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_RECEIVE_NO_BUF: %u",
+			    rndis.rx_no_buf);
+		net_buf_add_le32(buf, rndis.rx_no_buf);
+		break;
+
+		/* IEEE 802.3 */
+	case RNDIS_OBJECT_ID_802_3_PERMANENT_ADDRESS:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_802_3_PERMANENT_ADDRESS");
+		memcpy(net_buf_add(buf, sizeof(rndis.mac)), rndis.mac,
+		       sizeof(rndis.mac));
+		break;
+	case RNDIS_OBJECT_ID_802_3_CURR_ADDRESS:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_802_3_CURR_ADDRESS");
+		memcpy(net_buf_add(buf, sizeof(rndis.mac)), rndis.mac,
+		       sizeof(rndis.mac));
+		break;
+	case RNDIS_OBJECT_ID_802_3_MCAST_LIST:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_802_3_MCAST_LIST");
+		net_buf_add_le32(buf, 0xE0000000); /* 224.0.0.0 */
+		break;
+	case RNDIS_OBJECT_ID_802_3_MAX_LIST_SIZE:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_802_3_MAX_LIST_SIZE");
+		net_buf_add_le32(buf, 1); /* one address */
+		break;
+
+		/* Vendor information */
+	case RNDIS_OBJECT_ID_GEN_VENDOR_ID:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_VENDOR_ID");
+		net_buf_add_le32(buf, CONFIG_USB_DEVICE_VID);
+		break;
+	case RNDIS_OBJECT_ID_GEN_VENDOR_DESC:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_VENDOR_DESC");
+		memcpy(net_buf_add(buf, sizeof(manufacturer) - 1), manufacturer,
+		       sizeof(manufacturer) - 1);
+		break;
+	case RNDIS_OBJECT_ID_GEN_VENDOR_DRV_VER:
+		SYS_LOG_DBG("RNDIS_OBJECT_ID_GEN_VENDOR_DRV_VER");
+		net_buf_add_le32(buf, drv_version);
+		break;
+	default:
+		SYS_LOG_WRN("Unhandled query for Object ID 0x%x", object_id);
+		break;
+	}
+
+	buf_len = buf->len - sizeof(*rsp);
+
+	if (buf_len) {
+		rsp->status = sys_cpu_to_le32(RNDIS_CMD_STATUS_SUCCESS);
+	} else {
+		rsp->status = sys_cpu_to_le32(RNDIS_CMD_STATUS_NOT_SUPP);
+	}
+
+	/* Can be zero if object_id not handled / found */
+	rsp->buf_len = sys_cpu_to_le32(buf_len);
+
+	rsp->len = sys_cpu_to_le32(buf_len + sizeof(*rsp));
+
+	SYS_LOG_DBG("buf_len %u rsp->len %u buf->len %u",
+		    buf_len, rsp->len, buf->len);
+
+	rndis_queue_rsp(buf);
+
+	/* Nofity about ready reply */
+	rndis_notify_rsp();
+
+	return 0;
+}
+
+static int rndis_set_handle(u8_t *data, u32_t len)
+{
+	struct rndis_set_cmd *cmd = (void *)data;
+	struct rndis_set_cmd_complete *rsp;
+	struct net_buf *buf;
+	u32_t object_id;
+	u8_t *param;
+
+	buf = net_buf_alloc(&rndis_tx_pool, K_NO_WAIT);
+	if (!buf) {
+		SYS_LOG_ERR("Cannot get free buffer");
+		return -ENOMEM;
+	}
+
+	object_id = sys_le32_to_cpu(cmd->object_id);
+
+	SYS_LOG_DBG("req_id 0x%x Object ID 0x%x buf_len %u buf_offset %u",
+		    sys_le32_to_cpu(cmd->req_id), object_id,
+		    sys_le32_to_cpu(cmd->buf_len),
+		    sys_le32_to_cpu(cmd->buf_offset));
+
+	rsp = net_buf_add(buf, sizeof(*rsp));
+	rsp->type = sys_cpu_to_le32(RNDIS_CMD_SET_COMPLETE);
+	rsp->len = sys_cpu_to_le32(sizeof(*rsp));
+	rsp->req_id = cmd->req_id; /* same endianness */
+
+	/* Parameter starts at the beginning of req_id field ;) */
+	param = (u8_t *)&cmd->req_id + sys_le32_to_cpu(cmd->buf_offset);
```

Comment:
`cmd->buf_offset` also comes directly from the USB peer and it's used as an offset here.  This allows arbitrary memory reads.

Reply:
I will fix code to add check that param is not going out of the range in this case.